### PR TITLE
Implement touch API

### PR DESCRIPTION
**Pull request by jamesmcgill**
### Description

Implement legacy Touch API using new input system

### Changes made

- Populate touch functions
- Add some basic tests

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Assets/Tests/InputSystem/CoreTests_API_Touch.cs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            first.azimuthAngle == second.azimuthAngle &&
+            first.radius == second.radius &&
+            first.radiusVariance == second.radius;
+    }
+
+    [UnityTest]
+    [Category("API")]
+    public IEnumerator API_CanReadTouchEventsThroughTouchesAPI()
+    {
+        Assert.That(Input.touchSupported, Is.False);
```
Comment (Rene-Damm) [inline Assets/Tests/InputSystem/CoreTests_API_Touch.cs:36]: Prob good to try the other related APIs here before Touchscreen is added to make sure we get the expected defaults.

File: Packages/com.unity.input/Runtime/Input.cs [hunk 2]
```diff
@@ -647,27 +648,125 @@ public static bool GetKeyUp(KeyCode key)
...
+            {
+                case InputSystem.TouchPhase.Began: return TouchPhase.Began;
+                case InputSystem.TouchPhase.Moved: return TouchPhase.Moved;
+                case InputSystem.TouchPhase.Ended: return TouchPhase.Ended;
+                case InputSystem.TouchPhase.Canceled: return TouchPhase.Canceled;
+                case InputSystem.TouchPhase.Stationary: return TouchPhase.Stationary;
+                    //InputSystem.TouchPhase.None  // These are filtered out and shouldn't be seen
+            }
+            throw new ArgumentOutOfRangeException(nameof(phase), $"Not expected phase value: {phase}");
+        }
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input.cs]: Makes me wonder whether we should go and unify the two. We could add `None` to the trunk side version and kill the `InputSystem` side `TouchPhase`. Binary breaking change but just seems confusing we not only have two such enums now but are actively using both.

Anyway, not directly related to this PR. Just something for us to consider, I think.

File: Assets/Tests/InputSystem/CoreTests_API_Touch.cs (hunk 1 above)
Comment (Rene-Damm) [reply Assets/Tests/InputSystem/CoreTests_API_Touch.cs:36]: Also, test out of range access with `GetTouch`.

File: Packages/com.unity.input/Runtime/Input.cs [hunk 3]
```diff
@@ -647,27 +648,125 @@ public static bool GetKeyUp(KeyCode key)
...
+            touch.radiusVariance = variance;
+
+            return touch;
         }
 
         public static Touch GetTouch(int index)
         {
-            return default;
+            if (index >= touchCount)
+                throw new IndexOutOfRangeException();
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input.cs]: `IndexOutOfRangeException` should not be thrown this way. Use `ArgumentOutOfRangeException` instead.

Comment (Rene-Damm) [reply Packages/com.unity.input/Runtime/Input.cs]: (also need to make sure we are throwing a type equivalent to what the old implementation throws; can be subtype, though)

File: Packages/com.unity.input/Runtime/Input.cs [hunk 4]
```diff
@@ -647,27 +648,125 @@ public static bool GetKeyUp(KeyCode key)
...
+            }
+        }
+        public static Touch[] touches
+        {
+            get
+            {
+                var result = new List<Touch>();
+                if (Touchscreen.current != null)
+                {
+                    var convertedTouches = from t in Touchscreen.current.touches
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input.cs]: This needs to use the `EnhancedTouch` API. Polling `Touchscreen` directly will not give the right results. It'll drop touches and lead to incorrect touch sequences.

File: Packages/com.unity.input/Runtime/Input.cs (hunk 3 above)
Comment (jamesmcgill) [reply Packages/com.unity.input/Runtime/Input.cs]: I changed it to match what I found here: https://github.cds.internal.unity3d.com/unity/unity/blob/ccbb0cecf33d617817244907b804def62b34e6b9/Runtime/Input/InputBindings.h

I.e. ArgumentException("Invalid touch index.");

Comment (jamesmcgill) [reply Packages/com.unity.input/Runtime/Input.cs]: Sorry I mean: throw new ArgumentOutOfRangeException("Invalid touch index.");

Comment (Rene-Damm) [reply Packages/com.unity.input/Runtime/Input.cs]: First arg must be parameter name.

```
throw new ArgumentOutOfRangeException(nameof(index));
```

File: Packages/com.unity.input/Runtime/Input_Touch.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                EnhancedTouch.Touch.s_GlobalState.playerState.UpdateActiveTouches();
+                return EnhancedTouch.Touch.s_GlobalState.playerState.activeTouchCount;
+            }
+        }
+        public static Touch[] touches
+        {
+            get
+            {
+                if (!EnhancedTouchSupport.enabled)
+                    EnhancedTouchSupport.Enable();
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input_Touch.cs]: This needs to be enabled in `OnEnteringPlayMode`.

File: Packages/com.unity.input/Runtime/Input_Touch.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                throw new ArgumentOutOfRangeException(nameof(index), "Invalid touch index.");
+
+            return ToTouch(EnhancedTouch.Touch.activeTouches[index]);
+        }
+
+        public static int touchCount
+        {
+            get
+            {
+                EnhancedTouch.Touch.s_GlobalState.playerState.UpdateActiveTouches();
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input_Touch.cs]: Why this one?

File: Packages/com.unity.input/Runtime/Input_Touch.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+                return EnhancedTouch.Touch.s_GlobalState.playerState.activeTouchCount;
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input_Touch.cs]: This should use `Touch.activeTouches.Count`.

Comment (jamesmcgill) [reply Packages/com.unity.input/Runtime/Input_Touch.cs]: I'm currently seeing a memory leak somewhere with this function. 
I found that `Touch.activeTouches.Count` was calling `UpdateActiveTouches()` and then creating a temporary List, then returning the size of the list.
It actually has nothing to do with `UpdateActiveTouches()` creating a temporary list, but it was my first thought. I left it here because I thought it would be cheaper, but yeah, this is using the insides directly. I will change it back to use the list Count.

File: Packages/com.unity.input/Runtime/Input_Touch.cs (hunk 6 above)
Comment (jamesmcgill) [reply Packages/com.unity.input/Runtime/Input_Touch.cs]: See below (was needed to mimic `Touch.activeTouches.Count` behaviour)

File: Packages/com.unity.input/Runtime/Input.cs [hunk 8]
```diff
@@ -66,6 +67,10 @@ public static partial class Input // Partial to allow us to inject project-speci
 
         internal static void OnEnteringPlayMode()
         {
+            // EnhancedTouch is required for the Touch API
+            if (!EnhancedTouchSupport.enabled)
+                EnhancedTouchSupport.Enable();
```
Comment (Rene-Damm) [inline Packages/com.unity.input/Runtime/Input.cs]: ATM this thing functions like a counter. So just calling Enable here and Disable in Reset prob makes sense.
